istioctl/cmd/describe.go

```diff
@@ -1186,3 +1197,94 @@ func containerReady(pod *v1.Pod, containerName string) (bool, error) {
 	}
 	return false, fmt.Errorf("no container %q in pod", containerName)
 }
+
+// describePeerAuthentication fetches all PeerAuthentication in workload and root namespace.
+// It lists the ones applied to the pod, and the current active mTLS mode.
+// When the client doesn't have access to root namespace, it will only show workload namespace Peerauthentications.
+func describePeerAuthentication(writer io.Writer, configClient istioclient.Interface, istioNamespace, workloadNamespace string, pod *v1.Pod, podsLabels k8s_labels.Set) error { // nolint: lll
+	workloadPAList, err := configClient.SecurityV1beta1().PeerAuthentications(workloadNamespace).List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to fetch workload namespace PeerAuthentication: %v", err)
+	}
+
+	hasRootNamespaceAccess := true
+	rootPAList, err := configClient.SecurityV1beta1().PeerAuthentications(istioNamespace).List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		hasRootNamespaceAccess = false
+		fmt.Fprintf(writer, "Cannot analyze active mTLS mode since failed to fetch root namespace PeerAuthentication: %v:\n", err)
+	}
+
+	allPAs := append(rootPAList.Items, workloadPAList.Items...)
+	if len(allPAs) == 0 {
+		return nil
+	}
+
+	var cfgs []*config.Config
+	for _, pa := range allPAs {
+		pa := pa
+		cfg := crdclient.TranslateObject(&pa, config.GroupVersionKind(pa.GroupVersionKind()), "")
+		cfgs = append(cfgs, &cfg)
+	}
+
+	matchedPA := findMatchedConfigs(podsLabels, cfgs)
+	printConfigs(writer, matchedPA)
+
+	if hasRootNamespaceAccess {
+		activePA := authnv1beta1.ComposePeerAuthentication(istioNamespace, matchedPA)
+		printPeerAuthentication(writer, activePA)
+	}
+
+	return nil
+}
+
+// Workloader is used for matching all configs
+type Workloader interface {
+	GetSelector() *typev1beta1.WorkloadSelector
+}
+
+// findMatchedConfigs should filter out unrelated configs that are not matched given podsLabels.
+// Configs passed into this method should only contains workload's namespaces configs
+// and rootNamespaces configs, caller should be responsible for controlling configs passed
+// in.
+func findMatchedConfigs(podsLabels k8s_labels.Set, configs []*config.Config) []*config.Config {
+	var cfgs []*config.Config
+
+	for _, cfg := range configs {
+		cfg := cfg
+		spec := cfg.Spec.(Workloader)
+		if spec.GetSelector() == nil || len(spec.GetSelector().MatchLabels) == 0 {
+			cfgs = append(cfgs, cfg)
+		} else {
+			selector := spec.GetSelector().GetMatchLabels()
+			slr := k8s_labels.SelectorFromSet(selector)
```

Comment:
you do not need the if-else check, `k8s_labels.SelectorFromSet(nil)` will return a selector that matches everything, see the comment of the function.

Reply:
How about the edge case of `len(spec.GetSelector().MatchLabels) == 0`, is that a valid case?

Reply:
I think it also handles this, you can double check the function comment and test it a little bit.

Reply:
I've tested in cluster and that works, but there is no 0 size labels map.
Don't think in normal case, there is a 0 size map, it normally nil or size > 0. 
However, not sure whether there is such a case.

Reply:
I'm not sure what you mean, are you saying the `k8s_labels.SelectorFromSet()` can not handle 0 size map?

The function comment clearly says that `A nil and empty Sets are considered equivalent to Everything().` so it should handle 0 size map without problems.

```
// SelectorFromSet returns a Selector which will match exactly the given Set. A
// nil and empty Sets are considered equivalent to Everything().
// It does not perform any validation, which means the server will reject
// the request if the Set contains invalid values.
func SelectorFromSet(ls Set) Selector {
```

Reply:
Ah that's right, cool will do that.